### ISSUE TITLE
Add Distributed Cache

### ISSUE THREAD

**Issue by StephanEwen** [stratosphere/stratosphere#506]
Add a Hadoop-like Distributed Cache utility.

The distributed cache is needed for tasks that require files to be accessible with local file paths. Files are registered at the Program level with (file, name). The utility must take the files and write them to the local temp directory, assigning them a random file name and make the path available under the name.

**Comment by StephanEwen** [comment]
The distributed cache would be usable as follows, from the perspective of the someone who writes a Stratosphere program:
1. The files of interest must be stored in a place that can be accessed from all workers (most commonly HDFS)
2. When submitting a job, you register the files that you need locally cached at the job. For the new java API, that would be the `ExecutionEnvironment`, for the old api the `Plan`. The method could be called `registerCachedFile(filePath, name)`.
3. Within a task (let's say in the `open()` method) , you access the cache to give you the local path of that file `File f = getRuntimeContext().getDistributedCache().getFile(name)`

Under the hood, we need a the following (here is a suggestion, feel free to suggest improvements):
- Registered cached files are written into the job configuration (the same way as the job-wide switch to activate/deactivate the accumulators works)
- The TaskManager has a component called `FileCache`. When a task is deployed to the TaskManager, it checks that job configuration for registered (filePath, name) entries.
- The FileCache is invoked with that path and name and the job id (to qualify the name by job). If that file is not locally cached, it copies the file to the temp directory (same temp directory as the IOManager uses)
- When a task is unregistered, the cache should remove the tmp file. Given that another task from the same job may be registered shortly after, I suggest to have a 5 second delay (or so) before clearing files, to prevent unnecessary repeated downloads from HDFS to the local tmp directory.

**Comment by fhueske** [comment]
That would introduce a hard dependency on an HDFS setup when ever you want to use the Distributed Cache feature.

We should at least add an exception for Stratosphere's local mode, and maybe later also for setups with no HDFS (although these should be rather rare), for example via shared dirs such as NFS.

**Comment by StephanEwen** [comment]
I don't think it needs. If we add files with the normal URIs  (hdfs:///some/file/) or (file:///another/file), we can use the `FileSystem` api to copy files. That should transparently handle it, independent of HDFS, no?

**Comment by fhueske** [comment]
Sure, but somewhere the file system needs to be chosen.
All I am saying is, that we should not rely on HDFS for the local case and add support for other file systems in distributed environments later.

**Comment by fhueske** [comment]
Eh, OK. I got it wrong...

Proposal sounds good to me.
